src/read/elf/version.rs

```diff
@@ -293,6 +293,134 @@ impl<'data, Elf: FileHeader> VerdauxIterator<'data, Elf> {
+        )>,
+    > {
+        if self.data.is_empty() {
+            return Ok(None);
+        }
+
+        // Don't move self.data so that we can simply add section_length
+        // to it to move to the next vendor section
+        let mut data = self.data;
+        let section_length = *data
+            .read::<endian::U32<Elf::Endian>>()
+            .read_error("ELF GNU attributes vendor section is too short")?;
+        let vendor_name = data
+            .read_string()
+            .read_error("ELF GNU attributes vendor section is too short")?;
+
+        let tags = TagIterator::new(
+            self.endian,
+            data.read_bytes(
+                // calculate the remaining space in this section
+                // section - size field - vendor name string - NULL terminator
+                section_length.get(self.endian) as usize
+                    - core::mem::size_of::<endian::U32<Elf::Endian>>()
+                    - vendor_name.len()
+                    - 1,
+            )
+            .read_error("ELF GNU attributes vendor section is too short")?,
+        );
+
+        let next = section_length.get(self.endian);
+        if next != 0 {
+            self.data
+                .skip(next as usize)
+                .read_error("Invalid ELF vd_next")?;
+        } else {
+            self.data = Bytes(&[]);
+        }
+        Ok(Some((section_length, vendor_name, tags)))
+    }
+}
+
+/// An iterator over the attribute tags in a GNU attributes section
+#[derive(Debug, Clone)]
+pub struct TagIterator<'data, Elf: FileHeader> {
+    endian: Elf::Endian,
+    data: Bytes<'data>,
+    should_exit: bool,
+}
+
+impl<'data, Elf: FileHeader> TagIterator<'data, Elf> {
+    pub(super) fn new(endian: Elf::Endian, data: Bytes<'data>) -> Self {
+        TagIterator {
+            endian,
+            data: data,
+            should_exit: false,
+        }
+    }
+
+    /// Return the next tag.
+    ///
+    /// The format of attributes looks like this:
+    /// ```text
+    /// [ <file-tag> <size> <attribute>*
+    /// | <section-tag> <size> <section-number>* 0 <attribute>*
+    /// | <symbol-tag> <size> <symbol-number>* 0 <attribute>*
+    /// ]+
+    /// ```
+    /// At the moment only file tags are supported, other types will report an error.
+    ///
+    /// This iterator returns the (tag, data) pair, allowing the user to access the raw data for
+    /// the tags. The data is an array of attributes, each of which is an attribute tag folowed by
+    /// either a uleb128 encoded integer or a NULL terminated string
+    pub fn next(&mut self) -> Result<Option<(u8, &'data [u8])>> {
+        if self.should_exit || self.data.is_empty() {
+            return Ok(None);
+        }
+
+        let mut tag_data = self.data;
+        let tag = tag_data
+            .read::<u8>()
+            .read_error("GNU Attributes tag not correctly sized")?;
+
+        // If this is not a file tag, bail
+        if *tag != 1 {
```

Comment:
Why do we need this check? We don't parse the contents of the sub-subsection anywhere, so it doesn't matter if it contains ULEB128s. Even the file sub-subsection will contain ULEB128s because attribute tags are ULEB128s.